sets/maps.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sets
+
+// IntersectMapKeys finds the intersection of all m keys, where intersection is
+// defined as keys that exist in all m. It always returns an allocated map, even
+// if the intersection is the empty set.
+//
+// It does not modify any of the given inputs, but also does not deep copy any
+// values. That means the returned map may have keys and values that point to
+// the same objects as in the original map.
+func IntersectMapKeys[K comparable, V any](maps ...map[K]V) map[K]V {
+	if len(maps) == 0 {
+		return make(map[K]V)
+	}
+
+	// Pre-compute the maximum possible allocation to minimize allocs. Here we
+	// find the map with the least number of elements and allocate a map of that
+	// size, since we know that is the maximum possible intersection size.
+	var smallestIdx int
+	for i, m := range maps {
+		// Short-circuit: if any of the maps are the empty set, we know the
+		// intersection is the empty set.
+		if len(m) == 0 {
+			return make(map[K]V)
+		}
+
+		if len(m) < len(maps[smallestIdx]) {
+			smallestIdx = i
+		}
+	}
+	smallestMap := maps[smallestIdx]
+	final := make(map[K]V, len(smallestMap))
+	for k, v := range smallestMap {
+		final[k] = v
+	}
+
+	// Compute the intersection.
+	for i, m := range maps {
+		// Short-circuit: we've already got the smallest possible intersection
+		// (empty set), so there's no point in continuing.
+		if len(final) == 0 {
+			return final
+		}
```

Comment:
nit: should this be moved outside the loop, it feels strange to be in here when you dont need to come into the loop for this.

Reply:
We want to check on each iteration because lower on ~L68, we're deleting entries from the map that don't exist in the slice. So then when we go to process the next slice, we want to check again if the intersection is empty. We're modifying `final` in place.

Reply:
ack